lib/pos_consensus_controller.go

```diff
@@ -28,13 +29,50 @@ func NewConsensusController(params *DeSoParams, blockchain *Blockchain, mempool
 	}
 }
 
-func (cc *ConsensusController) Init() {
-	// This initializes the FastHotStuffEventLoop based on the blockchain state. This should
-	// only be called once the blockchain has synced, the node is ready to join the validator
-	// network, and the node is able validate blocks in the steady state.
-	//
-	// TODO: Implement this later once the Blockchain struct changes are merged. We need to be
-	// able to fetch the tip block and current persisted view from DB from the Blockchain struct.
+// ConsensusController.Start initializes and starts the FastHotStuffEventLoop based on the
+// blockchain state. This should only be called once the blockchain has synced, the node is
+// ready to join the validator network, and the node is able validate blocks in the steady state.
+func (cc *ConsensusController) Start() error {
+	// Hold the write consensus controller's lock for thread-safety.
+	cc.lock.Lock()
+	defer cc.lock.Unlock()
+
+	// Hold the blockchain's read lock so that the chain cannot be mutated underneath us. In practice,
+	// this is a no-op, but it guarantees thread-safety in the event that other parts of the codebase
+	// change.
+	cc.blockchain.ChainLock.RLock()
+	defer cc.blockchain.ChainLock.RUnlock()
+
+	// Fetch the current tip of the chain
+	tipBlock := cc.blockchain.BlockTip()
+
+	// Fetch the validator set at each safe block
+	tipBlockWithValidators, err := cc.fetchValidatorListsForSafeBlocks([]*MsgDeSoHeader{tipBlock.Header})
+	if err != nil {
+		return errors.Errorf("ConsensusController.Start: Error fetching validator list for tip blocks: %v", err)
+	}
+
+	// Fetch the safe blocks that are eligible to be extended from by the next incoming tip block
+	safeBlocks, err := cc.blockchain.GetSafeBlocks()
```

Comment:
This reminds me that I should add a comment on `GetSafeBlocks` that the caller should acquire a read lock before calling.

Reply:
Done. I added a comment for it above the GetSafeBlocks function signature.